pkg/controllers/infracluster/azure.go

```diff
@@ -0,0 +1,278 @@
+func (r *InfraClusterController) getAzureLocation(ctx context.Context) (string, error) {
+	// Devise Azure location via MAPI providerSpec.
+	machineSpec, err := getAzureMAPIProviderSpec(ctx, r.Client)
+	if err != nil || machineSpec == nil {
+		return "", errUnableToGetAzureMAPIProviderSpec
+	}
+
+	if machineSpec.Location == "" {
+		return "", fmt.Errorf("unable to get Azure Location: %w", err)
+	}
+
+	return machineSpec.Location, nil
+}
+
+func createNewAzureSecret(ctx context.Context, cl client.Client, azureClientSecret []byte) error {
+	azureSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      clusterSecretName,
+			Namespace: defaultCAPINamespace,
+		},
+		Immutable: ptr.To(true),
+		Data: map[string][]byte{
+			"clientSecret": azureClientSecret,
+		},
+	}
+
+	if err := cl.Create(ctx, azureSecret); err != nil && !cerrors.IsAlreadyExists(err) {
+		return fmt.Errorf("failed to create new Azure Cluster Secret: %w", err)
+	}
+
+	return nil
+}
+
+// ensureClusterSecret ensures the AzureClusterSecret exists, and if it doesn't, creates it.
+func (r *InfraClusterController) ensureClusterSecret(ctx context.Context, capzManagerBootstrapSecret corev1.Secret) error {
+	// Get the Azure Cluster Secret.
+	// CAPZ controllers expect a Secret Ref with a different data structure to what the secret we get from the Cluster Credential Operator above provides.
+	// That is why we are creating a new secret below to copy the values to.
+	clusterSecret := &corev1.Secret{}
+	if err := r.Get(ctx, client.ObjectKey{Namespace: defaultCAPINamespace, Name: clusterSecretName}, clusterSecret); err != nil {
+		if !cerrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get Azure Cluster Secret: %w", err)
+		}
+
+		// The Azure Cluster Secret does not exist, so it needs to be created.
+		if err := createNewAzureSecret(ctx, r.Client, capzManagerBootstrapSecret.Data["azure_client_secret"]); err != nil {
+			return fmt.Errorf("failed to create Azure Cluster secret: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// ensureClusterIdentity ensures the AzureClusterSecret exists, and if it doesn't, creates it.
+func (r *InfraClusterController) ensureClusterIdentity(ctx context.Context, capzManagerBootstrapSecret corev1.Secret) error {
+	azureClusterIdentity := &azurev1.AzureClusterIdentity{}
+	// Get the Azure Cluster Identity.
+	if err := r.Get(ctx, client.ObjectKey{Namespace: defaultCAPINamespace, Name: r.Infra.Status.InfrastructureName}, azureClusterIdentity); err != nil {
+		if !cerrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get Azure Cluster Identity: %w", err)
+		}
+
+		azureClientID, ok := capzManagerBootstrapSecret.Data["azure_client_id"]
+		if !ok {
+			return fmt.Errorf("failed to get azureClientID: %w", err)
+		}
+
+		azureTenantID, ok := capzManagerBootstrapSecret.Data["azure_tenant_id"]
+		if !ok {
+			return fmt.Errorf("failed to get azureTenantID: %w", err)
+		}
+
+		azureClusterIdentity = &azurev1.AzureClusterIdentity{
```

Comment:
Could instantiate this object in a helper to reduce the complexity of this function